### PR TITLE
Implement signing and signature verification.

### PR DESCRIPTION
**Pull request by BobWall23**
Added GPG signature packet after the literal data packet in the encrypted data in ironsftp secure files. When decrypting, validate the signature.

### CODE REVIEW

File: iron/gpg-packet.c [hunk 1]
```diff
@@ -0,0 +1,1150 @@
...
+    /* The hash is computed over the entire public key packet, the entire UID packet (except the stupid length is
+     * expanded to four bytes), and the first part of the signature packet. It ends up with a trailer that is the
+     * version, 0xff, and the four-byte length of the hashed data, MSB first.
+     */
+
+    /* First, generate the tag/length for the public key packet, hash that, then hash the packet contents. */
+    int len = generate_gpg_tag_and_size(pubkey_pkt->tag, pubkey_pkt->len, buf);
+    SHA256_Update(&ctx, buf, len);
+    SHA256_Update(&ctx, sshbuf_ptr(pubkey_pkt->data), sshbuf_len(pubkey_pkt->data));
+
```
Comment (zmre) [inline iron/gpg-packet.c:548]: Does the second SHA256_Update not overwrite the &ctx? Does it append results? So this is the sha of the length concatenated with the sha of the data?

File: iron/gpg-packet.c [hunk 2]
```diff
@@ -0,0 +1,1150 @@
...
+    /* Next, do the same with the UID packet, if there is one. Need to fiddle with the tag/length, because when
+     * GPG hashes it, it expands the length out to four bytes instead of one. Grr.
+     */
+    if (uid_pkt != NULL) {
+        buf[0] = 0xb4;   //  uid_pkt->tag converted to a tag byte
```
Comment (zmre) [inline iron/gpg-packet.c]: 0xb4 -- constant?

File: iron/gpg-packet.c (hunk 1 above)
Comment (BobWall23) [inline iron/gpg-packet.c:548]: It's the SHA of (length concatenated with data).

File: iron/gpg-packet.c [hunk 3]
```diff
@@ -0,0 +1,1150 @@
...
+    sshbuf_put_u32(body, iron_gpg_now());
+    sshbuf_put_u8(body, 5);         //  Length of key lifetime subpacket
+    sshbuf_put_u8(body, GPG_SIG_SUBPKT_KEY_LIFETIME);
+    sshbuf_put_u32(body, 0);            //  Does not expire
+    sshbuf_put_u8(body, 5);         //  Length of preferred symmetric algorithm subpacket
+    sshbuf_put_u8(body, GPG_SIG_SUBPKT_PREF_SYM_ALGO);
+    sshbuf_put_u8(body, GPG_SKALGO_AES256);
+    sshbuf_put_u8(body, GPG_SKALGO_AES192);
+    sshbuf_put_u8(body, GPG_SKALGO_AES128);
+    sshbuf_put_u8(body, 0);         //  No fourth preferred SK algorithm
```
Comment (zmre) [inline iron/gpg-packet.c:485]: So we put three different algorithm identifiers one after the next?

Comment (BobWall23) [inline iron/gpg-packet.c:485]: Yes - it's a fixed array of up to four preferred algorithms. I think there are similar arrays for PK algorithms and hash algorithms, and maybe for compression algorithms.

File: iron/gpg-packet.c [hunk 4]
```diff
@@ -0,0 +1,1150 @@
...
+extract_gpg_one_pass_signature_packet(const u_char * buf, int buf_len, u_char * key_id)
+{
+	if (buf_len < ONE_PASS_SIG_LEN) return -1;
+
+	const u_char * ptr = buf;
+	if (*(ptr++) != GPG_OPS_VERSION) return -2;
+	if (*(ptr++) != GPG_SIGCLASS_BINARY_DOC) return -3;
+	if (*(ptr++) != GPG_HASHALGO_SHA256) return -4;
+	if (*(ptr++) != GPG_PKALGO_RSA_ES) return -5;
+	memcpy(key_id, ptr, GPG_KEY_ID_LEN);
```
Comment (zmre) [inline iron/gpg-packet.c:958]: What are all these return codes? Maybe use constants?

Comment (BobWall23) [inline iron/gpg-packet.c:958]: We really want a file full of constants that no one is expected to do anything with? They are negative numbers that are different so I could tell where it was jumping out while troubleshooting.  I could just change them all to -1.

File: iron/gpg-key.c [hunk 5]
```diff
@@ -182,6 +182,36 @@ clamp_and_reverse_seckey(u_char * sk)
...
+ *
+ *  The GPG keygrip is a shortened representation (i.e. hash) of the parameters of the public key. The hash
+ *  is just SHA1. The RSA keygrip is so much simpler than the curve25519 one - an RSA keygrip is just a SHA1
+ *  hash of the public key parameter n.
+ *
+ *  @param key RSA key (only needs public params populated)
+ *  @param grip Place to write keygrip. At least SHA_DIGEST_LENGTH bytes.
+ */
+void
+generate_gpg_rsa_keygrip(const Key * rsa_key, u_char * grip)
```
Comment (zmre) [inline iron/gpg-key.c:195]: test

Comment (BobWall23) [reply iron/gpg-key.c:195]: Right back atcha.

**Pull request merged**